File: controllers/messageController.js
const pool = require('../config/db');

// GET /api/messages/conversations
exports.getConversations = async (req, res) => {
  try {
    const userId = req.user.id;

    const [conversations] = await pool.query(
      `SELECT 
        CASE WHEN m.sender_id = ? THEN m.receiver_id ELSE m.sender_id END as other_user_id,
        u.name as other_user_name, u.role as other_user_role, u.profile_photo,
        m.content as last_message,
        m.created_at as last_message_time,
        (SELECT COUNT(*) FROM messages m2 WHERE m2.sender_id = CASE WHEN m.sender_id = ? THEN m.receiver_id ELSE m.sender_id END AND m2.receiver_id = ? AND m2.is_read = 0) as unread_count
       FROM messages m
       JOIN users u ON u.id = CASE WHEN m.sender_id = ? THEN m.receiver_id ELSE m.sender_id END
       WHERE m.id IN (
         SELECT MAX(id) FROM messages
         WHERE sender_id = ? OR receiver_id = ?
         GROUP BY LEAST(sender_id, receiver_id), GREATEST(sender_id, receiver_id)
       )
       ORDER BY m.created_at DESC`,
      [userId, userId, userId, userId, userId, userId]
    );

    res.json({ conversations });
  } catch (err) {
    console.error('Get conversations error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/messages/:userId
exports.getMessages = async (req, res) => {
  try {
    const userId = req.user.id;
    const otherUserId = req.params.userId;

    const [messages] = await pool.query(
      `SELECT m.*, u.name as sender_name
       FROM messages m
       JOIN users u ON m.sender_id = u.id
       WHERE (m.sender_id = ? AND m.receiver_id = ?) OR (m.sender_id = ? AND m.receiver_id = ?)
       ORDER BY m.created_at ASC`,
      [userId, otherUserId, otherUserId, userId]
    );

    // Mark messages as read
    await pool.query(
      'UPDATE messages SET is_read = 1 WHERE sender_id = ? AND receiver_id = ? AND is_read = 0',
      [otherUserId, userId]
    );

    res.json({ messages });
  } catch (err) {
    console.error('Get messages error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// POST /api/messages
exports.sendMessage = async (req, res) => {
  try {
    const { receiverId, content } = req.body;
    const senderId = req.user.id;

    // Check receiver exists
    const [users] = await pool.query('SELECT id FROM users WHERE id = ?', [receiverId]);
    if (users.length === 0) return res.status(404).json({ error: 'Recipient not found.' });

    const [result] = await pool.query(
      'INSERT INTO messages (sender_id, receiver_id, content) VALUES (?, ?, ?)',
      [senderId, receiverId, content]
    );

    res.status(201).json({
      message: 'Message sent.',
      data: {
        id: result.insertId,
        sender_id: senderId,
        receiver_id: receiverId,
        content,
        sender_name: req.user.name,
        is_read: 0,
        created_at: new Date()
      }
    });
  } catch (err) {
    console.error('Send message error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

File: backend-spring/src/main/java/com/agrivalue/controller/MessageController.java
package com.agrivalue.controller;

import com.agrivalue.entity.Message;
import com.agrivalue.repository.MessageRepository;
import com.agrivalue.repository.UserRepository;
import com.agrivalue.security.JwtUserPrincipal;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/messages")
public class MessageController {

    private final MessageRepository messageRepo;
    private final UserRepository userRepo;

    public MessageController(MessageRepository messageRepo, UserRepository userRepo) {
        this.messageRepo = messageRepo;
        this.userRepo = userRepo;
    }

    @GetMapping("/conversations")
    public ResponseEntity<?> getConversations(@AuthenticationPrincipal JwtUserPrincipal principal) {
        List<Message> latestMessages = messageRepo.findLatestConversations(principal.getId());

        List<Map<String, Object>> conversations = latestMessages.stream().map(m -> {
            Integer otherUserId = m.getSenderId().equals(principal.getId()) ? m.getReceiverId() : m.getSenderId();
            Map<String, Object> conv = new LinkedHashMap<>();
            conv.put("other_user_id", otherUserId);
            conv.put("last_message", m.getContent());
            conv.put("last_message_time", m.getCreatedAt());
            conv.put("unread_count", messageRepo.countUnread(otherUserId, principal.getId()));
            userRepo.findById(otherUserId).ifPresent(u -> {
                conv.put("other_user_name", u.getName());
                conv.put("other_user_role", u.getRole().name());
                conv.put("profile_photo", u.getProfilePhoto());
            });
            return conv;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("conversations", conversations));
    }

    @GetMapping("/{userId}")
    @Transactional
    public ResponseEntity<?> getMessages(@AuthenticationPrincipal JwtUserPrincipal principal,
                                          @PathVariable Integer userId) {
        List<Message> messages = messageRepo.findConversation(principal.getId(), userId);
        messageRepo.markAsRead(userId, principal.getId());

        List<Map<String, Object>> result = messages.stream().map(m -> {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("id", m.getId());
            map.put("sender_id", m.getSenderId());
            map.put("receiver_id", m.getReceiverId());
            map.put("content", m.getContent());
            map.put("is_read", m.getIsRead());
            map.put("created_at", m.getCreatedAt());
            userRepo.findById(m.getSenderId()).ifPresent(u -> map.put("sender_name", u.getName()));
            return map;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("messages", result));
    }

    @PostMapping
    public ResponseEntity<?> sendMessage(@AuthenticationPrincipal JwtUserPrincipal principal,
                                          @RequestBody Map<String, Object> body) {
        Integer receiverId = Integer.valueOf(body.get("receiverId").toString());
        String content = body.get("content").toString();

        if (userRepo.findById(receiverId).isEmpty()) {
            return ResponseEntity.notFound().build();
        }

        Message message = Message.builder()
                .senderId(principal.getId()).receiverId(receiverId).content(content).build();
        message = messageRepo.save(message);

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("id", message.getId());
        data.put("sender_id", principal.getId());
        data.put("receiver_id", receiverId);
        data.put("content", content);
        data.put("sender_name", principal.getName());
        data.put("is_read", false);
        data.put("created_at", message.getCreatedAt());

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of("message", "Message sent.", "data", data));
    }
}
